eWonConnector-gateway/src/main/java/com/hms_networks/americas/sc/ignition/threading/PollingThread.java

```diff
@@ -0,0 +1,353 @@
+   * @since 1.0.0
+   */
+  public void stopGracefully() {
+    // Cancel the thread gracefully
+    boolean mayInterruptIfRunning = false;
+    stop(mayInterruptIfRunning);
+  }
+
+  /**
+   * Stops the polling thread immediately. In-progress polls are interrupted.
+   *
+   * @throws IllegalStateException if the polling thread executor service has been shutdown.
+   * @since 1.0.0
+   */
+  public void stopImmediately() {
+    // Cancel the thread immediately
+    boolean mayInterruptIfRunning = true;
+    stop(mayInterruptIfRunning);
+  }
+
+  /**
+   * Stops the polling thread using the specified interrupt behavior, and shuts down the executor
+   * service.
+   *
+   * <p>This method should be called when the thread is no longer needed, as it will prevent the
+   * thread from being restarted.
+   *
+   * @param mayInterruptIfRunning true if in-progress polls should be interrupted; otherwise,
+   *     in-progress polls are allowed to complete
+   * @throws IllegalStateException if the polling thread executor service has already been shutdown.
+   * @since 1.0.0
+   */
+  public void shutdown(boolean mayInterruptIfRunning) {
+    // Check if the executor service is shutdown
+    if (executorService.isShutdown()) {
+      throw new IllegalStateException("Polling thread executor service is already shutdown.");
+    }
+
+    // Shutdown the thread (if running)
+    if (isRunning()) {
+      stop(mayInterruptIfRunning);
+    }
+
+    // Shutdown the executor service
+    executorService.shutdown();
+  }
+
+  /**
+   * Stops the polling thread gracefully, and shuts down the executor service. In-progress polls are
+   * allowed to complete.
+   *
+   * <p>This method should be called when the thread is no longer needed, as it will prevent the
+   * thread from being restarted.
+   *
+   * @throws IllegalStateException if the polling thread executor service has already been shutdown.
+   * @since 1.0.0
+   */
+  public void shutdownGracefully() {
+    // Shutdown the thread gracefully
+    boolean mayInterruptIfRunning = false;
+    shutdown(mayInterruptIfRunning);
+  }
+
+  /**
+   * Stops the polling thread immediately, and shuts down the executor service. In-progress polls
+   * are interrupted.
+   *
+   * <p>This method should be called when the thread is no longer needed, as it will prevent the
+   * thread from being restarted.
+   *
+   * @throws IllegalStateException if the polling thread executor service has already been shutdown.
+   * @since 1.0.0
+   */
+  public void shutdownImmediately() {
+    // Shutdown the thread immediately
+    boolean mayInterruptIfRunning = true;
+    shutdown(mayInterruptIfRunning);
+  }
+
+  /**
+   * Gets the thread running status.
+   *
+   * @return The thread running status.
+   * @throws IllegalStateException if the polling thread executor service has been shutdown.
+   * @since 1.0.0
+   */
+  public boolean isRunning() {
```

Comment:
This returns false if canceled, but throws an exception if done. That is a little inconsistent. Perhaps change to be consistent.

Reply:
The idea is that the PollingThread can be stopped and started without shutting down the underlying executor.

The shutdown() method/API should only be invoked as cleanup or pre-disposal of the pollingthread, thus an exception is thrown if shutdown() has already been called.

If the thread is simply stopped though, isRunning() will return false, and the thread can be started again.

Reply:
Ah. I was confusing the thread with the executor service.